mesa/agent.py

```diff
@@ -348,29 +348,103 @@ def agg(self, attribute: str, func: Callable) -> Any:
+        default_value: Any = None,
+    ) -> list[Any] | list[list[Any]]:
         """
         Retrieve the specified attribute(s) from each agent in the AgentSet.
 
         Args:
             attr_names (str | list[str]): The name(s) of the attribute(s) to retrieve from each agent.
+            handle_missing (str, optional): How to handle missing attributes. Can be
+
+                                            -'error' (default), raises an AttributeError if attribute is missing.
+                                            -'skip' which skips the agents missing the attribute.
+                                            'default' to return the specified `default_value`.
+
+            default_value (Any, optional): Used in conjunction with ``handle_missing='default``
+                                     The default value to return if 'handle_missing' is set to 'default'
+                                     and the agent does not have the attribute.
 
         Returns:
-            list[Any]: A list with the attribute value for each agent in the set if attr_names is a str
-            list[list[Any]]: A list with a list of attribute values for each agent in the set if attr_names is a list of str
+            list[Any]: A list of attribute values for each agent if attr_names is a str.
+            list[list[Any]]: A list of lists of attribute values for each agent if attr_names is a list of str.
 
         Raises:
-            AttributeError if an agent does not have the specified attribute(s)
+            AttributeError: If 'handle_missing' is 'error' and the agent does not have the specified attribute(s).
+        """
+
+        # Check if the attr_names is a single string or a list of attributes.
+        is_single_attr = isinstance(attr_names, str)
+
+        # Branch earlier based on the `handle_missing` option to avoid repeated checks
+        match handle_missing:
+            case "error":
+                if is_single_attr:
+                    return [
+                        self._get_or_raise(agent, attr_names) for agent in self._agents
+                    ]
+                else:
+                    return [
+                        [self._get_or_raise(agent, attr) for attr in attr_names]
+                        for agent in self._agents
+                    ]
+
+            case "default":
+                if is_single_attr:
+                    return [
+                        self._get_with_default(agent, attr_names, default_value)
+                        for agent in self._agents
+                    ]
+                else:
+                    return [
+                        [
+                            self._get_with_default(agent, attr, default_value)
+                            for attr in attr_names
+                        ]
+                        for agent in self._agents
+                    ]
+
+            case "skip":
+                if is_single_attr:
+                    return [
```

Comment:
I think we need an actual for loop here to properly skip errors, so something like this:
```python
values = []
for agent in self._agents:
    try:
        values.append(getattr(agent, attr_names))
    except AttributeError:
        continue
return values
```

Reply:
I think you are indeed correct. I'll add a test for this corner case as well.